### PR TITLE
feat(wallet): improve displaying assets

### LINKED ISSUE #3900: Add shared logic for asset pagination
In the wallet, there's a fixed `50` assets to be fetched, and the pagination could be done, but we don't have any logic implemented. As the way of fetching the assets is similar in wallet and wallet-dashboard, we could create a shared hook or logic to paginate these assets.

https://github.com/iotaledger/iota/blob/55938963d5b7f1bbc4ba79ec97e571998a9927aa/apps/wallet/src/ui/app/hooks/useGetNFTs.ts#L32-L38

### LINKED ISSUE #3938: Improve paginator in assets page
## Description

Improve paginator in assets page
https://github.com/iotaledger/iota/pull/3758/files#r1821264610

Depends on https://github.com/iotaledger/iota/issues/3900

### PR DESCRIPTION
**Pull request by brancoder**
## Links to any relevant issues

resolves #3900 
resolves #3938

### CODE REVIEW

**Review by cpl121 — CHANGES_REQUESTED** [review]

File: apps/wallet-dashboard/app/(protected)/assets/page.tsx [hunk 1]
```diff
@@ -27,33 +24,30 @@ const ASSET_CATEGORIES: { label: string; value: AssetCategory }[] = [
...
-
-    const { data: ownedObjects } = data || {};
+    const { data, isFetching, fetchNextPage, hasNextPage } = useGetOwnedObjects(
+        account?.address,
+        undefined,
+        50,
+    );
 
-    const [visual, nonVisual] = (() => {
+    const [visual, nonVisual] = useMemo(() => {
```
Comment (marc2332) [inline apps/wallet-dashboard/app/(protected)/assets/page.tsx]: Unnecessary useMemo

**Review by marc2332 — CHANGES_REQUESTED** [review]

File: apps/wallet-dashboard/app/(protected)/assets/page.tsx [hunk 2]
```diff
@@ -27,30 +24,28 @@ const ASSET_CATEGORIES: { label: string; value: AssetCategory }[] = [
...
     const account = useCurrentAccount();
-    const ownedObjectsQuery = useGetOwnedObjects(account?.address, undefined, limit);
-
-    const { data, pagination } = useCursorPagination(ownedObjectsQuery);
-
-    const { data: ownedObjects } = data || {};
+    const { data, isFetching, fetchNextPage, hasNextPage } = useGetOwnedObjects(
+        account?.address,
+        undefined,
+        50,
```
Comment (cpl121) [inline apps/wallet-dashboard/app/(protected)/assets/page.tsx]: we can add a constant for this hardcoded value

File: apps/wallet-dashboard/components/AssetsList.tsx [hunk 3]
```diff
@@ -16,12 +22,35 @@ const ASSET_LAYOUT: Record<AssetCategory, string> = {
...
+            fetchNextPage();
+        }
+    }, [isIntersecting, fetchNextPage, hasNextPage, isFetchingNextPage]);
+
     return (
         <div className={ASSET_LAYOUT[selectedCategory]}>
             {assets.map((asset) => (
                 <AssetTileLink key={asset.digest} asset={asset} type={selectedCategory} />
             ))}
+            <div ref={observerElem}>
```
Comment (cpl121) [inline apps/wallet-dashboard/components/AssetsList.tsx:48]: wouldnt you use scroll for both screens? imo, we can include the scroll in the other assets category

image
image

**Review by cpl121 — CHANGES_REQUESTED** [review]

**Review by cpl121 — APPROVED** [review]
NJ!

File: apps/wallet-dashboard/components/AssetsList.tsx [hunk 4]
```diff
@@ -4,10 +4,17 @@
...
+import { useEffect, useRef } from 'react';
+import { useOnScreen } from '@iota/core';
+import cl from 'clsx';
 
 interface AssetListProps {
     assets: IotaObjectData[];
     selectedCategory: AssetCategory;
+    hasNextPage?: boolean;
+    isFetchingNextPage?: boolean;
+    fetchNextPage?: () => void;
```
Comment (marc2332) [inline apps/wallet-dashboard/components/AssetsList.tsx]: These are never optional
```suggestion
    hasNextPage: boolean;
    isFetchingNextPage: boolean;
    fetchNextPage: () => void;
```

File: apps/wallet-dashboard/app/(protected)/assets/page.tsx
Comment (marc2332) [inline apps/wallet-dashboard/app/(protected)/assets/page.tsx:1]: You could also optimize the code a bit so only the selected assets of the category are mapped

(I also simplified the code)
```ts
	const { data, isFetching, fetchNextPage, hasNextPage } = useGetOwnedObjects(
        account?.address,
        undefined,
        OBJECTS_PER_REQ,
    );

    const assets: IotaObjectData[] = [];

    for(const page of data?.pages || []){
        for(const asset of page.data){
            if (asset.data && asset.data.objectId) {
                if (selectedCategory == AssetCategory.Visual) {
                    if(hasDisplayData(asset)) {
                        assets.push(asset.data);
                    }
                } else if(selectedCategory == AssetCategory.Other) {
                    assets.push(asset.data);
                }
            }
        }
    }

    return (
        <Panel>
            <Title title="Assets" size={TitleSize.Medium} />
            <div className="px-lg">
                <div className="flex flex-row items-center justify-start gap-xs py-xs">
                    {ASSET_CATEGORIES.map((tab) => (
                        <Chip
                            key={tab.label}
                            label={tab.label}
                            onClick={() => setSelectedCategory(tab.value)}
                            selected={selectedCategory === tab.value}
                        />
                    ))}
                </div>

                <AssetList
                    assets={assets}
                    selectedCategory={selectedCategory}
                    hasNextPage={hasNextPage}
                    isFetchingNextPage={isFetching}
                    fetchNextPage={fetchNextPage}
                />
            </div>
        </Panel>
    );
```

File: apps/wallet-dashboard/components/AssetsList.tsx [hunk 5]
```diff
@@ -16,12 +23,35 @@ const ASSET_LAYOUT: Record<AssetCategory, string> = {
     [AssetCategory.Other]: 'flex flex-col overflow-auto py-sm',
 };
 
-export function AssetList({ assets, selectedCategory }: AssetListProps): React.JSX.Element {
+export function AssetList({
+    assets,
+    selectedCategory,
+    hasNextPage = false,
+    isFetchingNextPage = false,
```
Comment (marc2332) [inline apps/wallet-dashboard/components/AssetsList.tsx]:
```suggestion
    hasNextPage,
    isFetchingNextPage,
```

**Review by marc2332 — CHANGES_REQUESTED** [review]

**Review by panteleymonchuk — APPROVED** [review]

**Review by marc2332 — APPROVED** [review]

**Pull request merged**